### PR TITLE
clp-s: Add support for wildcard queries using "?" on a variable field.

### PR DESCRIPTION
**Pull request by wraymo**
# References

#392 

# Description

This PR adds the feature requested in #392. Now we can use `?` to match an arbitrary single character on a variable field. However, there are still two limitations.
+ Queries with a `?` in the middle return empty results. This should be fixes by #407.
+ Queries combined with `?`(`*`) and `\?`(`\*`) cannot work well. This will be fixed later.

# Validation performed

Compressed the logs below
```
{"a": "Hello world?", "b": "wwoorr\lldd", "c":"*world"}
{"a": "Hello *", "b": "?world"}
```
The queries below all returned correct results
```
b: \?world
b: ?world
b: ?wo?ld
b: wwoorr\\lldd
c: ?world
c: \*world
```

### CODE REVIEW

File: components/core/src/clp_s/search/StringLiteral.hpp [hunk 1]
```diff
@@ -68,7 +68,7 @@ class StringLiteral : public Literal {
             m_string_type = LiteralType::VarStringT;
         }
 
-        if (m_v.find('*') != std::string::npos) {
+        if (m_v.find('*') != std::string::npos || m_v.find('?') != std::string::npos) {
```
Comment (gibber9809) [inline components/core/src/clp_s/search/StringLiteral.hpp:71]: I know it's wrong in the original code as well, but could we fix this to handle escaped `*` and `?` properly as well? This should really be checking just for unescaped `*` or `?`, and ignoring them when they're escaped. E.g. `var\?` is just a variable string.

File: components/core/src/clp_s/search/Output.cpp [hunk 2]
```diff
@@ -915,7 +915,9 @@ void Output::populate_string_queries(std::shared_ptr<Expression> const& expr) {
             }
 
             std::unordered_set<int64_t>& matching_vars = m_string_var_match_map[query_string];
-            if (query_string.find('*') == std::string::npos) {
+            if (query_string.find('*') == std::string::npos
+                && query_string.find('?') == std::string::npos)
```
Comment (gibber9809) [inline components/core/src/clp_s/search/Output.cpp:919]: Similar to my other comment we should be handling escaping here as well. Specifically if there are no unescaped `*` or `?` we should enter this code block, process the string to unescape escaped characters as necessary, then do the precise search.

E.g. `\\var\?` should enter this code block and get unescaped to become `\var?` before performing exact match.

**Review by gibber9809 — CHANGES_REQUESTED** [review]

File: components/core/src/clp_s/search/Output.cpp (hunk 2 above)
Comment (wraymo) [reply components/core/src/clp_s/search/Output.cpp:919]: I think the function `wildcard_match_unsafe_case_sensitive` can handle escape characters correctly. The corner case was mentioned above and will be fixed later.

File: components/core/src/clp_s/search/StringLiteral.hpp (hunk 1 above)
Comment (wraymo) [reply components/core/src/clp_s/search/StringLiteral.hpp]: Right

File: components/core/src/clp_s/search/Output.cpp (hunk 2 above)
Comment (gibber9809) [reply components/core/src/clp_s/search/Output.cpp:919]: Right, but this code block uses `m_var_dict->get_entry_matching_value()` which doesn't respect escape characters.

Comment (gibber9809) [reply components/core/src/clp_s/search/Output.cpp:919]: Maybe for the condition here we can add a method like
```
bool has_unescaped_wildchards(std::string const& query) {
    bool escaped{false};
    for (size_t i = 0; i < query.size(); ++i) {
        if (query[i] == '*' || query[i] == '?') {
            return true;
        }
        if (query[i] == '\') {
            ++i;
        }
    }
}
```

and replace the if condition with `false == has_unescaped_wildcards(query)`?

**Review by gibber9809 — APPROVED** [review]
LGTM. Maybe just add (fixes #392) to the commit message -- besides that title is good as commit message.

**Comment by kirkrodrigues** [comment]
How about:

clp-s: Add support for using "?" in wildcard queries on single-word values (fixes #392).

**Pull request merged**